implementations/rust/ockam/ockam_api/src/nodes/service/forwarder.rs

```diff
@@ -1,44 +1,89 @@
+    DeleteSecureChannelRequest,
+};
 use crate::nodes::NodeManager;
+use crate::session::Session;
+
+const MAX_RECOVERY_TIME: Duration = Duration::from_secs(10);
+const MAX_CONNECT_TIME: Duration = Duration::from_secs(5);
 
 impl NodeManager {
     pub(super) async fn create_forwarder(
         &mut self,
         ctx: &mut Context,
-        req: &Request<'_>,
+        rid: Id,
         dec: &mut Decoder<'_>,
     ) -> Result<Vec<u8>> {
-        let CreateForwarder {
-            address,
-            alias,
-            at_rust_node,
-            ..
-        } = dec.decode()?;
-        let addr = MultiAddr::try_from(address.0.as_ref()).map_err(map_multiaddr_err)?;
-        let route = crate::multiaddr_to_route(&addr)
-            .ok_or_else(|| ApiError::generic("Invalid Multiaddr"))?;
-        debug!(%addr, ?alias, "Handling CreateForwarder request");
-
-        let forwarder = match alias {
+        let req: CreateForwarder = dec.decode()?;
+
+        debug!(addr = %req.address, alias = ?req.alias, "Handling CreateForwarder request");
+
+        let forwarder = match req.alias {
             Some(alias) => {
-                if at_rust_node {
-                    RemoteForwarder::create_static_without_heartbeats(ctx, route, alias.to_string())
-                        .await
+                let auth = Arc::new(req.identities);
+                let addr = connect(self, &req.address, &auth, req.mode).await?;
+                let alias = alias.to_string();
+                let fwdr = if req.at_rust_node {
+                    let route = addr.either(Route::from, identity);
+                    RemoteForwarder::create_static_without_heartbeats(ctx, route, alias).await
                 } else {
-                    RemoteForwarder::create_static(ctx, route, alias.to_string()).await
-                }
+                    // If connect returned an address it is a secure channel to an api node and
+                    // automatic recovery of a remote forwarder is enabled.
+                    match addr {
+                        Either::Left(a) => {
+                            let r = Route::from(a.clone());
+                            let f = RemoteForwarder::create_static(ctx, r, alias.clone()).await;
+                            if f.is_ok() {
+                                let c = Arc::new(ctx.async_try_clone().await?);
+                                let mut s = Session::new(a);
+                                let this = self.address.clone();
+                                enable_recovery(
+                                    &mut s,
+                                    this,
+                                    c,
+                                    req.address,
+                                    alias.to_string(),
+                                    auth,
+                                    req.mode,
+                                );
+                                self.sessions.lock().unwrap().add(s);
```

Comment:
I've seen that you always unwrap the `lock()` result. Is this done on purpose to let the mutex in a "poisoned" state if something panics?

Reply:
It is sort of the default thing to do unless one is sure that no invariants are broken if a panic occurs while holding the lock. While panics are not expected when dealing with sessions they are propagated across thread boundaries this way should they arise.